Mathlib/Combinatorics/SetFamily/Shatter.lean

```diff
@@ -0,0 +1,195 @@
+
+@[simp] lemma shatterer_eq : 𝒜.shatterer = 𝒜 ↔ IsLowerSet (𝒜 : Set (Finset α)) := by
+  refine ⟨fun h ↦ ?_, fun h ↦ Subset.antisymm (fun s hs ↦ ?_) $ subset_shatterer h⟩
+  · rw [←h]
+    exact isLowerSet_shatterer _
+  · obtain ⟨t, ht, hst⟩ := (mem_shatterer.1 hs).exists_superset
+    exact h hst ht
+
+@[simp] lemma shatterer_idem : 𝒜.shatterer.shatterer = 𝒜.shatterer := by simp
+
+@[simp] lemma shatter_shatterer : 𝒜.shatterer.Shatter s ↔ 𝒜.Shatter s := by
+  simp_rw [←mem_shatterer, shatterer_idem]
+
+protected alias ⟨_, Shatter.shatterer⟩ := shatter_shatterer
+
+private lemma aux (h : ∀ t ∈ 𝒜, a ∉ t) (ht : 𝒜.Shatter t) : a ∉ t := by
+  obtain ⟨u, hu, htu⟩ := ht.exists_superset; exact not_mem_mono htu $ h u hu
+
+/-- Pajor's variant of the **Sauer-Shelah lemma**. -/
+lemma card_le_card_shatterer (𝒜 : Finset (Finset α)) : 𝒜.card ≤ 𝒜.shatterer.card := by
+  refine memberFamily_induction_on 𝒜 ?_ ?_ ?_
+  · simp
+  · rfl
+  intros a 𝒜 ih₀ ih₁
+  set ℬ : Finset (Finset α) :=
+    ((memberSubfamily a 𝒜).shatterer ∩ (nonMemberSubfamily a 𝒜).shatterer).image (insert a)
+  have hℬ :
+    ℬ.card = ((memberSubfamily a 𝒜).shatterer ∩ (nonMemberSubfamily a 𝒜).shatterer).card
+  · refine card_image_of_injOn $ insert_erase_invOn.2.injOn.mono ?_
+    simp only [coe_inter, Set.subset_def, Set.mem_inter_iff, mem_coe, Set.mem_setOf_eq, and_imp,
+      mem_shatterer]
+    exact fun s _ ↦ aux (fun t ht ↦ (mem_filter.1 ht).2)
+  rw [←card_memberSubfamily_add_card_nonMemberSubfamily a]
+  refine (add_le_add ih₁ ih₀).trans ?_
+  rw [←card_union_add_card_inter, ←hℬ, ←card_disjoint_union]
+  swap
+  · simp only [disjoint_left, mem_union, mem_shatterer, mem_image, not_exists, not_and]
+    rintro _ (hs | hs) s - rfl
+    · exact aux (fun t ht ↦ (mem_memberSubfamily.1 ht).2) hs $ mem_insert_self _ _
+    · exact aux (fun t ht ↦ (mem_nonMemberSubfamily.1 ht).2) hs $ mem_insert_self _ _
+  refine card_mono $ union_subset (union_subset ?_ $ shatterer_mono $ filter_subset _ _) ?_
+  · simp only [subset_iff, mem_shatterer]
+    rintro s hs t ht
+    obtain ⟨u, hu, rfl⟩ := hs ht
+    rw [mem_memberSubfamily] at hu
+    refine ⟨insert a u, hu.1, inter_insert_of_not_mem fun ha ↦ ?_⟩
+    obtain ⟨v, hv, hsv⟩ := hs.exists_inter_eq_singleton ha
+    rw [mem_memberSubfamily] at hv
+    rw [←singleton_subset_iff (a := a), ←hsv] at hv
+    exact hv.2 $ inter_subset_right _ _
+  · refine forall_image.2 fun s hs ↦ mem_shatterer.2 fun t ht ↦ ?_
+    simp only [mem_inter, mem_shatterer] at hs
+    rw [subset_insert_iff] at ht
+    by_cases ha : a ∈ t
+    · obtain ⟨u, hu, hsu⟩ := hs.1 ht
+      rw [mem_memberSubfamily] at hu
+      refine ⟨_, hu.1, ?_⟩
+      rw [←insert_inter_distrib, hsu, insert_erase ha]
+    · obtain ⟨u, hu, hsu⟩ := hs.2 ht
+      rw [mem_nonMemberSubfamily] at hu
+      refine ⟨_, hu.1, ?_⟩
+      rwa [insert_inter_of_not_mem hu.2, hsu, erase_eq_self]
+
+protected lemma Shatter.of_compression (hs : (𝓓 a 𝒜).Shatter s) : 𝒜.Shatter s := by
```

Comment:
What's the purpose of marking this protected?

Reply:
Remnant from a time where it was (wrongly) called `Shatter.compression`